### PR TITLE
C++17

### PR DESCRIPTION
**Pull request by ObiWahn**
Finally enable C++17 for devel.

The old PR contains some information on Problems that had to be solved:
https://github.com/arangodb/arangodb/pull/7066

Latest PR runs:

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6620/
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6623/

### CODE REVIEW

File: 3rdParty/rocksdb/6.2/utilities/transactions/transaction_base.cc [hunk 1]
```diff
@@ -121,12 +121,48 @@ Status TransactionBaseImpl::TryLock(ColumnFamilyHandle* column_family,
                  assume_tracked);
 }
 
+#if defined(__APPLE__) && _LIBCPP_STD_VER > 14
+// miserable hack to access the container member of std::stack
```
Comment (ObiWahn) [inline 3rdParty/rocksdb/6.2/utilities/transactions/transaction_base.cc]: Try to remove this code! It should not be needed anymore.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6623/ - OK - except for known cluster error

**Comment by ObiWahn** [comment]
> * 3rdParty/V8/CMakeLists.txt doesn't switch the MSVS version properly for gyp
> * no gyp support for MSVS 2019 was added

Almost everything works with MSVS 2017 an upgrade is not urgent.

> * no patch descriptions in 3rdParty/README_maintainers.md was added.

There is an PR that will remove the mac-hack. And the other change to V8 was taken from
upstream. -> No patching required

**Comment by dothebart** [comment]
> > * 3rdParty/V8/CMakeLists.txt doesn't switch the MSVS version properly for gyp
> > * no gyp support for MSVS 2019 was added
> 
> Almost everything works with MSVS 2017 an upgrade is not urgent.

MSVS usually doesn't like switching compilers in the midst of a compile of one binary. Hence its all or nothing.

**Comment by ObiWahn** [comment]
80% of C++17 work with MSVC17. I think we should tackle the switch to 19 in another PR and fix the recommendation for developers. @dothebart @jsteemann what do you think?

https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019

**Comment by dothebart** [comment]
-  P0961R1 Relaxing the structured bindings customization point finding rules
 -  P0969R0 Allowing structured bindings to accessible members

?

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-matrix/2/

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6769/

File: CMakeLists.txt [hunk 2]
```diff
@@ -599,6 +609,8 @@ if (MSVC)
   add_definitions("-D_CRT_SECURE_NO_WARNINGS=1")
   add_definitions("-DFD_SETSIZE=8192")
   add_definitions("-DU_STATIC_IMPLEMENTATION=1")
+  add_definitions("-D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS=1") # try to remove this boost when
```
Comment (jsteemann) [inline CMakeLists.txt]:
```suggestion
  add_definitions("-D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS=1") # try to remove this when
```

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-mac-onlycatch/4733/

**Comment by KVS85** [comment]
Tests blue.

**Pull request merged**